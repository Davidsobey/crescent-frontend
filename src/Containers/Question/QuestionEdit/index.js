/**
 *
 * QuestionEdit
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { MenuItem } from 'material-ui/Menu';
import { CircularProgress } from 'material-ui/Progress';

import Select from '../../../Components/Select';
import Card from '../../../Components/Card';
import TextField from '../../../Components/TextField';
import Button from '../../../Components/Button';
import QuestionActions from '../../../Actions/QuestionActions';
import TestActions from '../../../Actions/TestActions';
import LinearProgress from '../../../Components/LinearProgress';

const required = value => value ? undefined : 'Required';
const number = value => value && isNaN(Number(value)) ? 'Must be a number' : undefined;

/* eslint-disable react/prefer-stateless-function */
class QuestionEdit extends React.Component {
  constructor(props) {
    super(props);
    this.submit = this.submit.bind(this);
  }

  componentWillMount() {
    this.props.dispatch(TestActions.getAll());
  }

  submit = (values) => {
    this.props.dispatch(QuestionActions.editQuestion(values));
  };

  render() {
    return (
      <Card width="600px" title="Edit Question">
        {this.props.question_loading ? (
          <div className="center">
            <CircularProgress color="secondary" />
          </div>
        ) : (
          <form
            onSubmit={this.props.handleSubmit(this.submit)}
            autoComplete="off"
            className="centerForm"
          >
            <div>
              {!this.props.tests_loading ? (
                <div>
                  <Field 
                    name="testId" 
                    label="Assessment Name" 
                    component={Select}
                    validate={[ required ]}
                  >
                    {(Array.isArray(this.props.tests) ? this.props.tests : [])
                    .map(test => (
                      <MenuItem value={test.id} key={test.id}>
                        {test.name}
                      </MenuItem>
                    ))}
                  </Field>
                </div>
              ) : (
                <div>
                  <LinearProgress color="secondary" />
                  Loading Tests
                </div>
              )}
              <div>
                <Field
                  name="title"
                  label="Question Name"
                  margin="normal"
                  component={TextField}
                  validate={[ required ]}
                />
              </div>
              <div>
                <Field
                  name="allocatedMarks"
                  label="Allocated Marks"
                  margin="normal"
                  component={TextField}
                  validate={[ required, number ]}
                />
              </div>
            </div>
            {this.props.question_editing ? (
              <div style={{width: '400px'}}>
                <LinearProgress color="secondary" />
                Editing Question
              </div>
            ) : (
              <div className="formAlignRight">
                <Button
                  className="buttonFormat"
                  variant="raised"
                  color="primary"
                  type="submit"
                >
                Edit Question
                </Button>
              </div>
            )}
          </form>
        )}
      </Card>
    );
  }
}
QuestionEdit.propTypes = {
  dispatch: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func,
  tests: PropTypes.array,
  tests_loading: PropTypes.bool,
  question_loading: PropTypes.bool,
  question_editing: PropTypes.bool,
};
const mapDispatchToProps = dispatch => ({
  dispatch,
});

// Decorate with reduxForm(). It will read the initialValues prop provided by connect()
const QuestionEdited = reduxForm({
  form: 'questionEdit', // a unique identifier for this form
})(QuestionEdit);

// You have to connect() to any reducers that you wish to connect to yourself
const FormState = connect(
  state => ({
    initialValues: state.QuestionReducer.question, // pull initial values from reducer
    tests: state.TestReducer.tests,
    tests_loading: state.TestReducer.loading,
    question_loading: state.QuestionReducer.question_loading,
    question_editing: state.QuestionReducer.question_editing,
  }),
  mapDispatchToProps,
)(QuestionEdited);

export default FormState;
