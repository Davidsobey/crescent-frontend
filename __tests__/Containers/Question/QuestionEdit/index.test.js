import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store'; // Smart components

// Component to be Questioned
import QuestionEdit from '../../../../src/Containers/Question/QuestionEdit/index.js';


const mockStore = configureStore();
const initialState = {
  TestReducer:{
    tests: [],
    loading: false,
  },
  QuestionReducer:{
    question: null,
    question_loading: false,
    question_editing: false,
  }
};
const store = mockStore(initialState);

describe('<QuestionEdit />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(<QuestionEdit store={store} />);
      const component = wrapper.dive();

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});