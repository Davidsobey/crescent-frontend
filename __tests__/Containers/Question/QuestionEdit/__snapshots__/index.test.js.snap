// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionEdit /> render() renders the component 1`] = `
<Connect(Form(QuestionEdit))
  destroyOnUnmount={true}
  dispatch={[Function]}
  enableReinitialize={false}
  forceUnregisterOnUnmount={false}
  form="questionEdit"
  getFormState={[Function]}
  initialValues={null}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  pure={true}
  question_editing={false}
  question_loading={false}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  tests={Array []}
  tests_loading={false}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
